thewalrus/fock_gradients.py

```diff
@@ -355,3 +359,101 @@ def grad_beamsplitter(T, theta, phi):  # pragma: no cover
+    for p, n in enumerate(pattern):
+        if n > 0:
+            yield p, dec(pattern, p)
+
+SQRT = np.sqrt(np.arange(1000))  # saving the time to recompute square roots
+
+def gaussian_gate(C, mu, Sigma, cutoff, num_modes, dtype=np.complex128): #pylint: disable=too-many-arguments
+    r"""Calculates the Fock representation of the gaussian gate.
+
+    Args:
+        C (complex): parameter for the gaussian gate
+        mu (vector[complex]): parameter for the gaussian gate
+        Sigma (array[complex]): parameter for the gaussian gate
+        cutoff (int): Fock ladder cutoff
+        num_modes (int): number of modes in the gaussian gate
+        dtype (data type): Specifies the data type used for the calculation
+
+    Returns:
+        array[complex]: The Fock representation of the gate
+    """
+    array = np.zeros(((cutoff,)*(2*num_modes)), dtype=dtype)
+    for n_current in range(1, 2*num_modes+1):
+        for idx in partition(num_modes, n_current, cutoff):
+            if idx == (0,)*(2*num_modes):
+                array[idx] = C
+            else:
+                array = fill_gaussian_gate_loop(array, idx, mu, Sigma)
+    return array
+
+@jit(nopython=True)
+def fill_gaussian_gate_loop(array, idx, mu, Sigma): # pragma: no cover #pylint: disable=undefined-loop-variable
+    "numba code to fill the gaussian gate according to the index given"
+    for i, val in enumerate(idx):
+        if val > 0:
+            break
+    ki = dec(idx, i)
+    u = mu[i] * array[ki]
+    for l, kl in remove(ki):
+        u -= SQRT[ki[l]] * Sigma[i, l] * array[kl]
+    array[idx] = u / SQRT[idx[i]]
+    return array
+
+def grad_gaussian_gate(gate, C, mu, Sigma, cutoff, num_modes, dtype=np.complex128): #pylint: disable=too-many-arguments
+    r"""Calculates the gradients of the gaussian gate.
+
+    Args:
+        gate (array[complex]): array representing the gate
+        C (complex): parameter for the gaussian gate
+        mu (vector[complex]): parameter for the gaussian gate
+        Sigma (array[complex]): parameter for the gaussian gate
+        cutoff (int): Fock ladder cutoff
+        num_modes (int): number of modes in the gaussian gate
+        dtype (data type): Specifies the data type used for the calculation
+
+    Returns:
+        array[complex]: The Fock representation of the gate
+    """
+    dG_dC = gate/C
+    dG_dmu = np.zeros_like(gate, dtype = dtype)
+    dG_dSigma = np.zeros_like(gate, dtype = dtype)
+    for n_current in range(1, 2*num_modes+1):
+        for idx in partition(num_modes, n_current, cutoff):
+            if not idx == (0,)*(len(gate.shape)):
+                dG_dmu, dG_dSigma = fill_grad_gaussian_gate_loop(dG_dmu, dG_dSigma, gate, idx, mu, Sigma)
+    return dG_dC, dG_dmu, dG_dSigma
+
+@jit(nopython=True)
+def fill_grad_gaussian_gate_loop(dG_dmu, dG_dSigma, gate, idx, mu, Sigma): # pragma: no cover #pylint: disable=undefined-loop-variable
```

Comment:
I suggest you find a way to solve the issue with variable `i` in line 451 other than by doing #pylint: disable=undefined-loop-variable

Reply:
Yes, I have several solutions and I will discuss them with Filippo.